### PR TITLE
Rewrite of iPXE chapter

### PR DESCRIPTION
**Pull request by lzap**
I am working with QA on a iPXE bug and I am finding this extremely confusing. I simplified it a ton.

First, I moved all the file copying and common actions onto a common prerequisite/precedure. This should be really here, I found cases when users were not told to copy this files.

Also, I am pushing more on using smart proxy as the default scenario, there are no longer examples using Foreman Server directly. This is more simple and it always works.

Then I corrected some confusing naming like "chainbooting" or "chainbooting Foreman". Technically, its just booting, but one particular case is when a server "boots PXELinux and it chanboots iPXE". Everything else should be just booting.

Cherry-pick into:

* [ ] Foreman 3.0
* [ ] Foreman 2.5 (Satellite 6.10)
* [ ] Foreman 2.4
* [ ] Foreman 2.3 (Satellite 6.9)

### CODE REVIEW

**Comment by lzap** [comment]
I noticed one thing, github renders asciidoc (sort of) so if you click on View File then you can also nicely see rendered version:

https://github.com/theforeman/foreman-documentation/blob/3d005d4242d21afad588137bcb6b720a0991ddb3/guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc

Cool.

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 1]
```diff
@@ -69,32 +50,76 @@ Before you begin, ensure that the following conditions are met:
...
+----
++
+. Optionally, configure Foreman discovery:
++
+* Install Foreman discovery image:
 +
-. In the {ProjectWebUI}, navigate to *Administer* > *Settings*, and click the *Provisioning* tab.
-. Locate the *Default PXE global template entry* row and in the *Value* column, change the value to *discovery*.
+[options="nowrap" subs="+quotes,attributes"]
+ifdef::satellite[]
```
Comment (maximiliankolb) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:111]: Is this ifdef=satellite really necessary? I think this also applies to Foreman and orcharhino.

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 2]
```diff
@@ -104,7 +129,8 @@ It has the following workflow:
 . Virtual machine starts
 . iPXE retrieves the network credentials using DHCP
 . iPXE retrieves the HTTP address using DHCP
-. iPXE chainloads the iPXE template from the template {SmartProxy}
+. iPXE loads the iPXE bootsrap template from the template {SmartProxy}
```
Comment (maximiliankolb) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:132]:
```suggestion
. iPXE loads the iPXE bootstrap template from {SmartProxy}
```

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 3]
```diff
@@ -104,7 +129,8 @@ It has the following workflow:
...
+. iPXE loads the iPXE template with MAC as a URL parameter from the template {SmartProxy}
```
Comment (maximiliankolb) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:133]:
```suggestion
. iPXE loads the iPXE template with MAC as a URL parameter from {SmartProxy}
```

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 4]
```diff
@@ -126,14 +152,6 @@ If you want to change the default values in the template, clone the template and
 
 .Procedure
```
Comment (maximiliankolb) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:154]:
```suggestion
```

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 5]
```diff
@@ -150,31 +168,25 @@ If you want to change the default values in the template, clone the template and
...
 }
 ----
 +
 If you use an isolated network, use a {SmartProxyServer} URL with TCP port `8000`, instead of the URL of {ProjectServer}.
 +
 [NOTE]
-Use `\http://{foreman-example-com}/unattended/iPXE?bootstrap=1` when {SmartProxy} HTTP endpoint is disabled (installer option --foreman-proxy-http false).
-Template {SmartProxy} plug-in has the default value `8000` when enabled and can be changed with `--foreman-proxy-http-port installer` option.
-In that case, use `\http://{smartproxy-example-com}:8000`.
+Use a different port than `8000` when changed with `--foreman-proxy-http-port installer` option.
```
Comment (maximiliankolb) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:183]:
```suggestion
If you have changed the port using the `--foreman-proxy-http-port installer` option, use your custom port.
```

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 6]
```diff
@@ -188,16 +200,12 @@ You can use this procedure only with bare metal hosts.
...
-
 .Configuring {ProjectName} Server to use iPXE
 
 You can use the default template to configure iPXE booting for hosts.
 If you want to change the default values in the template, clone the template and edit the clone.
 
-.Procedure
-
-. In the {ProjectWebUI}, navigate to *Hosts* > *Provisioning Templates*, enter `PXELinux chain iPXE` or, for BIOS systems, enter `PXELinux chain iPXE UNDI`, and click *Search*.
+. In the {ProjectWebUI}, navigate to *Hosts* > *Provisioning Templates*, enter `PXELinux chain iPXE` to use `ipxe.lkrn` or, for BIOS systems, enter `PXELinux chain iPXE UNDI` to use `undionly-ipxe.0`, and click *Search*.
```
Comment (maximiliankolb) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:208]:
```suggestion
. In the {ProjectWebUI}, navigate to *Hosts* > *Provisioning Templates*.
. Enter `PXELinux chain iPXE` to use `ipxe.lkrn` or, for BIOS systems, enter `PXELinux chain iPXE UNDI` to use `undionly-ipxe.0`, and click *Search*.
```

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 7]
```diff
@@ -223,65 +231,17 @@ If you want to change the default values in the template, clone the template and
...
 } else {
   filename "undionly.0";
 }
 ----
 +
-If you use an isolated network, use a {SmartProxyServer} URL with TCP port `8000`, instead of the URL of {ProjectServer}.
-+
 [NOTE]
-For `\http://{foreman-example-com}/unattended/iPXE`, you can also use a {ProjectName} {SmartProxy} `\http://{smartproxy-example-com}:8000/unattended/iPXE`.
+Use a different port than `8000` when changed with `--foreman-proxy-http-port installer` option.
```
Comment (maximiliankolb) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:246]: See above.

**Review by maximiliankolb — CHANGES_REQUESTED** [review]
Thanks for reworking this!

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 8]
```diff
@@ -69,32 +50,76 @@ Before you begin, ensure that the following conditions are met:
...
+----
+endif::[]
++
+* Copy the iPXE firmware with Linux kernel header to the TFTP directory:
++
+----
+# cp /usr/share/ipxe/ipxe.lkrn /var/lib/tftpboot/
+----
++
+* Copy the UNDI iPXE firmware to the TFTP directory:
```
[the remark this answers is not in the document]
Comment (lzap) [reply guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:100]: Explained.

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc (hunk 1 above)
Comment (lzap) [reply guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:111]: There is no RPM package upstream, let me do a xref instead.

**Comment by lzap** [comment]
I am amending all your changes @maximiliankolb thank you!

Also, Sayan from Red Hat did a test run on the current version of this PR and found some other issues which I am incorporating too. Quote from https://bugzilla.redhat.com/show_bug.cgi?id=1996519:

**SUCCESS - BIOS:**

1. BIOS system in libvirt
2. Selected "PXELinux chain iPXE" as the PXELinux template.  
3. Host powered up, chainbooted to discovery, discovered , custoimized and submitted for build.

```
   PXELoader : iPXE Embedded --> Works
   PXEloader : iPXE Chain BIOS --> Does not works and keeps looping
   PXEloader : None --> Works as well
```

I went through some googling and found that you have explicitly mentioned the fact that **make sure the PXE Loader is set to “None”** in the EFI workflow guide i.e. https://community.theforeman.org/t/discovery-ipxe-efi-workflow-in-foreman-1-20/13026 

But https://github.com/theforeman/foreman-documentation/blob/3d005d4242d21afad588137bcb6b720a0991ddb3/guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc is lacking that info if that is what needed to be done in case of both UEFI and BIOS 

DHCP section for BIOS:

```
if exists user-class and option user-class = "iPXE" {
  filename "http://192.168.126.2/unattended/iPXE?bootstrap=1";
} else {
  filename "undionly.0";
}
```

So now, as i mentioned earlier, "PXELinux chain iPXE UNDI" will try to load "undionly-ipxe.0" , none of the instruction says where to get this file from inside /var/lib/tftpboot . I am guessing we will need to just copy "/usr/share/ipxe/undionly.kpxe" by that name but need a confirmation on that as well in the doc.

**SUCCESS - UEFI:**

1. Copied /usr/share/ipxe/ipxe.efi to tftpboot
2. EFI system in libvirt
3. No changes made in any templates or template selections for OS.
4.. Host powered up, chainbooted to discovery, discovered , custoimized and submitted for build.

```
   PXELoader : iPXE Embedded --> Does not works
   PXEloader : iPXE Chain UEFI --> Does not works and keeps looping
   PXEloader : None --> Works Properly
```

DHCP section for UEFI:

```
if exists user-class and option user-class = "iPXE" {
  filename "http://satellite610.test.lan/unattended/iPXE?bootstrap=1";
} elsif option architecture = 00:06 {
  filename "ipxe.efi";
} elsif option architecture = 00:07 {
  filename "ipxe.efi";
} elsif option architecture = 00:09 {
  filename "ipxe.efi";
} else {
  filename "undionly.0";
}
```

In short, both are working i.e. UEFI as well as BIOS, and the common factor is "PXEloader" needs to be set to None in both of them.
So apart from the fact about "undionly-ipxe.0" loader, I very much suggest to mention the fact about selection of "PXEloader" to "None" [ or iPXE Embedded ] for BIOS based installation. [ For EFI it's already mentioned in your guide ].
Without the PXELoader fact mentioned, anyone can make the same mistake as mine here i.e. to select iPXE based Chain loaders in OS page and they will get stuck into that bootloop.

**Comment by lzap** [comment]
Also I am making sure the firmware filename is named correctly in Foreman core, it is a rare workflow but let's fix it: https://github.com/theforeman/foreman/pull/8800

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 9]
```diff
@@ -69,32 +49,68 @@ Before you begin, ensure that the following conditions are met:
...
+ifdef::satellite,orcharhino[]
+
+. Correct the SELinux file contexts:
++
+----
+# restorecon -RvF /var/lib/tftpboot/
 ----
+endif::[]
+
+* Copy the iPXE firmware with Linux kernel header to the TFTP directory:
```
Comment (melcorr) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:94]:
```suggestion
* Copy the iPXE firmware with the Linux kernel header to the TFTP directory:
```

**Review by melcorr — COMMENTED** [review]
nit but otherwise good to go.

**Comment by lzap** [comment]
Ok thanks all merging.
